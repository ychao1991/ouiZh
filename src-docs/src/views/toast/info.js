/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiToast } from '../../../../src/components';

export default () => (
  <OuiToast title="Icons should be rare" type="info" onClose={() => {}}>
    <p>
      Icons should be used rarely. They are good for warnings, but when paired
      with long titles they look out of place.
    </p>
  </OuiToast>
);
