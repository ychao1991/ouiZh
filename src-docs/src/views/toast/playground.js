/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { PropTypes } from 'react-view';
import { OuiToast } from '../../../../src/components/';
import {
  propUtilityForPlayground,
  iconValidator,
  createOptionalEnum,
  dummyFunction,
  simulateFunction,
} from '../../services/playground';

export default () => {
  const docgenInfo = Array.isArray(OuiToast.__docgenInfo)
    ? OuiToast.__docgenInfo[0]
    : OuiToast.__docgenInfo;
  const propsToUse = propUtilityForPlayground(docgenInfo.props);

  propsToUse.iconType = iconValidator(propsToUse.iconType);

  propsToUse.title = {
    ...propsToUse.title,
    type: PropTypes.String,
    value: 'Toast content',
  };

  propsToUse.color = createOptionalEnum(propsToUse.color);

  propsToUse.onClose = simulateFunction(propsToUse.onClose);

  return {
    config: {
      componentName: 'OuiToast',
      props: propsToUse,
      scope: {
        OuiToast,
      },
      imports: {
        '@opensearch-project/oui': {
          named: ['OuiToast'],
        },
      },
      customProps: {
        onClose: dummyFunction,
      },
    },
  };
};
