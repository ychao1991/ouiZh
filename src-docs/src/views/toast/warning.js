/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiToast } from '../../../../src/components';

export default () => (
  <OuiToast
    title="Sometimes a title is enough!"
    color="warning"
    iconType="help"
  />
);
