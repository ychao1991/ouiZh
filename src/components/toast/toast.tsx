/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  FunctionComponent,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';

import { CommonProps, keysOf } from '../common';
import { OuiScreenReaderOnly } from '../accessibility';
import { OuiI18n } from '../i18n';

import { IconType, OuiIcon } from '../icon';

import { OuiText } from '../text';

type ToastColor = 'primary' | 'success' | 'warning' | 'danger';

const colorToClassNameMap: { [color in ToastColor]: string } = {
  primary: 'ouiToast--primary',
  success: 'ouiToast--success',
  warning: 'ouiToast--warning',
  danger: 'ouiToast--danger',
};

export const COLORS = keysOf(colorToClassNameMap);

export interface OuiToastProps
  extends CommonProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  title?: ReactNode;
  color?: ToastColor;
  iconType?: IconType;
  onClose?: () => void;
}

export const OuiToast: FunctionComponent<OuiToastProps> = ({
  title,
  color,
  iconType,
  onClose,
  children,
  className,
  ...rest
}) => {
  const classes = classNames(
    'ouiToast',
    color ? colorToClassNameMap[color] : null,
    className
  );
  const headerClasses = classNames('ouiToastHeader', {
    'ouiToastHeader--withBody': children,
  });

  let headerIcon: ReactElement;

  if (iconType) {
    headerIcon = (
      <OuiIcon
        className="ouiToastHeader__icon"
        type={iconType}
        size="m"
        aria-hidden="true"
      />
    );
  }

  let closeButton;

  if (onClose) {
    closeButton = (
      <OuiI18n token="ouiToast.dismissToast" default="关闭通知">
        {(dismissToast: string) => (
          <button
            type="button"
            className="ouiToast__closeButton"
            aria-label={dismissToast}
            onClick={onClose}
            data-test-subj="toastCloseButton">
            <OuiIcon type="cross" size="m" aria-hidden="true" />
          </button>
        )}
      </OuiI18n>
    );
  }

  let optionalBody;

  if (children) {
    optionalBody = (
      <OuiText size="s" className="ouiToastBody">
        {children}
      </OuiText>
    );
  }

  return (
    <div className={classes} {...rest}>
      <OuiScreenReaderOnly>
        <p>
          <OuiI18n
            token="ouiToast.newNotification"
            default="出现新通知"
          />
        </p>
      </OuiScreenReaderOnly>

      <OuiI18n token="ouiToast.notification" default="通知">
        {(notification: string) => (
          <div
            className={headerClasses}
            aria-label={notification}
            data-test-subj="ouiToastHeader">
            {headerIcon}

            <span className="ouiToastHeader__title">{title}</span>
          </div>
        )}
      </OuiI18n>

      {closeButton}
      {optionalBody}
    </div>
  );
};
