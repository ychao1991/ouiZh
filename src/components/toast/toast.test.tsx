/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { findTestSubject, requiredProps } from '../../test';

import { COLORS, OuiToast } from './toast';

describe('OuiToast', () => {
  test('is rendered', () => {
    const component = render(
      <OuiToast {...requiredProps} title="test title">
        <p>Hi</p>
      </OuiToast>
    );

    expect(component).toMatchSnapshot();
  });

  describe('Props', () => {
    describe('title', () => {
      test('is rendered', () => {
        const component = <OuiToast title="toast title" />;
        expect(mount(component)).toMatchSnapshot();
      });
    });

    describe('color', () => {
      COLORS.forEach((color) => {
        test(`${color} is rendered`, () => {
          const component = <OuiToast color={color} title="test title" />;
          expect(mount(component)).toMatchSnapshot();
        });
      });
    });

    describe('iconType', () => {
      test('is rendered', () => {
        const component = <OuiToast iconType="user" title="test title" />;
        expect(mount(component)).toMatchSnapshot();
      });
    });

    describe('onClose', () => {
      test('is called when the close button is clicked', () => {
        const onCloseHandler = jest.fn();

        const component = mount(
          <OuiToast onClose={onCloseHandler} title="test title" />
        );
        const closeButton = findTestSubject(component, 'toastCloseButton');
        closeButton.simulate('click');

        expect(onCloseHandler).toBeCalledTimes(1);
      });
    });
  });
});
