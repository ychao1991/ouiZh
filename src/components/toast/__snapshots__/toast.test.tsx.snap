// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiToast Props color danger is rendered 1`] = `
<OuiToast
  color="danger"
  title="test title"
>
  <div
    className="ouiToast ouiToast--danger"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <span
          className="ouiToastHeader__title"
        >
          test title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast Props color primary is rendered 1`] = `
<OuiToast
  color="primary"
  title="test title"
>
  <div
    className="ouiToast ouiToast--primary"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <span
          className="ouiToastHeader__title"
        >
          test title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast Props color success is rendered 1`] = `
<OuiToast
  color="success"
  title="test title"
>
  <div
    className="ouiToast ouiToast--success"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <span
          className="ouiToastHeader__title"
        >
          test title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast Props color warning is rendered 1`] = `
<OuiToast
  color="warning"
  title="test title"
>
  <div
    className="ouiToast ouiToast--warning"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <span
          className="ouiToastHeader__title"
        >
          test title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast Props iconType is rendered 1`] = `
<OuiToast
  iconType="user"
  title="test title"
>
  <div
    className="ouiToast"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <OuiIcon
          aria-hidden="true"
          className="ouiToastHeader__icon"
          size="m"
          type="user"
        >
          <span
            aria-hidden="true"
            className="ouiToastHeader__icon"
            data-ouiicon-type="user"
            size="m"
          />
        </OuiIcon>
        <span
          className="ouiToastHeader__title"
        >
          test title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast Props title is rendered 1`] = `
<OuiToast
  title="toast title"
>
  <div
    className="ouiToast"
  >
    <OuiScreenReaderOnly>
      <p
        className="ouiScreenReaderOnly"
      >
        <OuiI18n
          default="A new notification appears"
          token="ouiToast.newNotification"
        >
          A new notification appears
        </OuiI18n>
      </p>
    </OuiScreenReaderOnly>
    <OuiI18n
      default="Notification"
      token="ouiToast.notification"
    >
      <div
        aria-label="Notification"
        className="ouiToastHeader"
        data-test-subj="ouiToastHeader"
      >
        <span
          className="ouiToastHeader__title"
        >
          toast title
        </span>
      </div>
    </OuiI18n>
  </div>
</OuiToast>
`;

exports[`OuiToast is rendered 1`] = `
<div
  aria-label="aria-label"
  class="ouiToast testClass1 testClass2"
  data-test-subj="test subject string"
>
  <p
    class="ouiScreenReaderOnly"
  >
    A new notification appears
  </p>
  <div
    aria-label="Notification"
    class="ouiToastHeader ouiToastHeader--withBody"
    data-test-subj="ouiToastHeader"
  >
    <span
      class="ouiToastHeader__title"
    >
      test title
    </span>
  </div>
  <div
    class="ouiText ouiText--small ouiToastBody"
  >
    <p>
      Hi
    </p>
  </div>
</div>
`;
